File: 백준/Silver/9012. 괄호/괄호.java
import java.io.*;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuffer sb = new StringBuffer();

        int N = Integer.parseInt(br.readLine());

        for(int i = 0; i < N; i++) {
            String[] line = br.readLine().split("");
            Stack<Boolean> stack = new Stack<>();

            boolean isGood = true;
            for(String s : line) {
                if(s.equals("(")) stack.push(true);
                else if(s.equals(")")) {
                    if(stack.isEmpty()) {
                        isGood = false;
                        break;
                    }
                    stack.pop();
                }
            }
            if(!stack.isEmpty()) isGood = false;
            if(isGood) sb.append("YES\n");
            else sb.append("NO\n");
        }
        System.out.println(sb);
    }
}

File: 백준/Silver/9012. 괄호/괄호.py
n = int(input())

rtn = ""
for i in range(n):
    check = True
    stack= []
    str = input()
    for s in str:
        if s == '(':
            stack.append("a")
        else:
            if len(stack) == 0:
                rtn += "NO\n"
                check = False
                break
            stack.pop()

    if not check:
        continue
    if len(stack) == 0:
        rtn += "YES\n"
    else:
        rtn += "NO\n"

print(rtn)
